### PR TITLE
libc/stream: Implement gets/puts for all streams

### PR DESCRIPTION
**Pull request by Gary-Hobson**
## Summary
Implement gets/puts for all streams

### CODE REVIEW

File: libs/libc/stream/lib_memsostream.c [hunk 1]
```diff
@@ -50,10 +50,33 @@ static void memsostream_putc(FAR struct lib_sostream_s *this, int ch)
       mthis->buffer[mthis->offset] = ch;
       mthis->offset++;
       this->nput++;
-      mthis->buffer[mthis->offset] = '\0';
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_memsostream.c:53]: why remove the tail zero? which is required to ensure the buffer always terminated with '\0'.

File: libs/libc/stream/lib_memsostream.c [hunk 2]
```diff
@@ -50,10 +50,33 @@ static void memsostream_putc(FAR struct lib_sostream_s *this, int ch)
...
+
+static int memsostream_puts(FAR struct lib_sostream_s *this,
+                            FAR const void *buf, int len)
+{
+  int ncopy;
+  FAR struct lib_memsostream_s *mthis = (FAR struct lib_memsostream_s *)this;
+
+  DEBUGASSERT(this);
+
+  ncopy = mthis->offset + len < mthis->buflen ? len :
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_memsostream.c:68]:
```suggestion
  ncopy = mthis->offset + len + 1 < mthis->buflen ? len :
            mthis->buflen - mthis->offset - 1;
```

File: libs/libc/stream/lib_memsostream.c [hunk 3]
```diff
@@ -50,10 +50,33 @@ static void memsostream_putc(FAR struct lib_sostream_s *this, int ch)
...
+          mthis->buflen - mthis->offset;
+  if (ncopy > 0)
+    {
+      memcpy(mthis->buffer + mthis->offset, buf, ncopy);
+      mthis->public.nput += ncopy;
+      mthis->offset += ncopy;
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_memsostream.c:74]: append '\0'

File: libs/libc/stream/lib_rawinstream.c [hunk 4]
```diff
@@ -66,6 +66,33 @@ static int rawinstream_getc(FAR struct lib_instream_s *this)
...
+  /* Attempt to read one character */
+
+  nread = _NX_READ(rthis->fd, buffer, len);
+  if (nread >= 0)
+    {
+      this->nget += nread;
+    }
+  else
+    {
+      nread = get_errno();
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_rawinstream.c:90]: nread = _NX_GETERRNO(nread);

File: libs/libc/stream/lib_rawsistream.c [hunk 5]
```diff
@@ -66,6 +66,33 @@ static int rawsistream_getc(FAR struct lib_sistream_s *this)
...
+  /* Attempt to read a buffer */
+
+  nread = _NX_READ(rthis->fd, buffer, len);
+  if (nread >= 0)
+    {
+      this->nget += nread;
+    }
+  else
+    {
+      nread = get_errno();
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_rawsistream.c:90]: nread = _NX_GETERRNO(nread);

File: libs/libc/stream/lib_rawsostream.c [hunk 6]
```diff
@@ -73,6 +73,45 @@ static void rawsostream_putc(FAR struct lib_sostream_s *this, int ch)
...
+       * was awakened by a signal.  Zero would not be a valid return value
+       * from _NX_WRITE().
+       */
+
+      errcode = _NX_GETERRNO(nwritten);
+      DEBUGASSERT(nwritten < 0);
+    }
+  while (errcode == EINTR);
+
+  return errcode;
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_rawsostream.c:112]: should we return nwritten in case of succeed.

File: libs/libc/stream/lib_stdoutstream.c [hunk 7]
```diff
@@ -64,6 +65,44 @@ static void stdoutstream_putc(FAR struct lib_outstream_s *this, int ch)
   while (get_errno() == EINTR);
 }
 
+/****************************************************************************
+ * Name: stdoutstream_puts
+ ****************************************************************************/
+
+static int stdoutstream_puts(FAR struct lib_outstream_s *this,
+                              FAR const void *buffer, int len)
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_stdoutstream.c:73]:
```suggestion
                             FAR const void *buffer, int len)
```

File: libs/libc/stream/lib_stdoutstream.c [hunk 8]
```diff
@@ -64,6 +65,44 @@ static void stdoutstream_putc(FAR struct lib_outstream_s *this, int ch)
...
+
+      errcode = get_errno();
+
+      /* EINTR (meaning that fputc was interrupted by a signal) is the only
+       * recoverable error.
+       */
+    }
+  while (errcode == EINTR);
+
+  return errcode;
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_stdoutstream.c:103]: return the written size when pass

File: libs/libc/stream/lib_stdsostream.c [hunk 9]
```diff
@@ -63,6 +63,43 @@ static void stdsostream_putc(FAR struct lib_sostream_s *this, int ch)
...
+
+      errcode = get_errno();
+
+      /* EINTR (meaning that fputc was interrupted by a signal) is the only
+       * recoverable error.
+       */
+    }
+  while (errcode == EINTR);
+
+  return errcode;
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_stdsostream.c:100]: return length when pass

File: libs/libc/stream/lib_stdsistream.c [hunk 10]
```diff
@@ -52,6 +52,31 @@ static int stdsistream_getc(FAR struct lib_sistream_s *this)
...
+{
+  FAR struct lib_stdsistream_s *sthis = (FAR struct lib_stdsistream_s *)this;
+  int nread = 0;
+  char *ret;
+
+  DEBUGASSERT(this);
+
+  /* Get the buffer from the incoming stream */
+
+  ret = fgets(buffer, len, sthis->stream);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_stdsistream.c:70]: why use fgets, but not fread

File: libs/libc/stream/lib_stdsostream.c [hunk 11]
```diff
@@ -63,6 +63,43 @@ static void stdsostream_putc(FAR struct lib_sostream_s *this, int ch)
   while (get_errno() == EINTR);
 }
 
+/****************************************************************************
+ * Name: stdsostream_puts
+ ****************************************************************************/
+
+static int stdsostream_puts(FAR struct lib_sostream_s *this,
+                             FAR const void *buffer, int len)
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_stdsostream.c:71]:
```suggestion
                            FAR const void *buffer, int len)
```

File: libs/libc/stream/lib_rawsostream.c [hunk 12]
```diff
@@ -73,6 +73,45 @@ static void rawsostream_putc(FAR struct lib_sostream_s *this, int ch)
   while (errcode == EINTR);
 }
 
+/****************************************************************************
+ * Name: rawsostream_puts
+ ****************************************************************************/
+
+static int rawsostream_puts(FAR struct lib_sostream_s *this,
+                             FAR const void *buffer, int len)
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_rawsostream.c:81]:
```suggestion
                            FAR const void *buffer, int len)
```

File: libs/libc/stream/lib_nulloutstream.c [hunk 13]
```diff
@@ -39,6 +39,16 @@ static void nulloutstream_putc(FAR struct lib_outstream_s *this, int ch)
   this->nput++;
 }
 
+static int nulloutstream_puts(FAR struct lib_outstream_s *this,
+                               FAR const void *buffer, int len)
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_nulloutstream.c:43]:
```suggestion
                              FAR const void *buffer, int len)
```

File: libs/libc/stream/lib_memsostream.c (hunk 1 above)
Comment (Gary-Hobson) [reply libs/libc/stream/lib_memsostream.c]: done

File: libs/libc/stream/lib_rawinstream.c (hunk 4 above)
Comment (Gary-Hobson) [reply libs/libc/stream/lib_rawinstream.c]: done

File: libs/libc/stream/lib_rawsostream.c [hunk 14]
```diff
@@ -73,6 +73,45 @@ static void rawsostream_putc(FAR struct lib_sostream_s *this, int ch)
...
+          this->nput += nwritten;
+          return nwritten;
+        }
+
+      /* The only expected error is EINTR, meaning that the write operation
+       * was awakened by a signal.  Zero would not be a valid return value
+       * from _NX_WRITE().
+       */
+
+      errcode = _NX_GETERRNO(nwritten);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_rawsostream.c:107]: nwritten = _NX_GETERRNO(nwritten);
and remove errcode

File: libs/libc/stream/lib_stdoutstream.c [hunk 15]
```diff
@@ -64,6 +65,44 @@ static void stdoutstream_putc(FAR struct lib_outstream_s *this, int ch)
...
+  do
+    {
+      result = fwrite(buffer, len, 1, sthis->stream);
+      if (result >= 0)
+        {
+          this->nput += result;
+          return result;
+        }
+
+      errcode = _NX_GETERRNO(result);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_stdoutstream.c:95]: result = _NX_GETERRNO(result);
and remove errcode

File: libs/libc/stream/lib_stdsostream.c [hunk 16]
```diff
@@ -63,6 +63,43 @@ static void stdsostream_putc(FAR struct lib_sostream_s *this, int ch)
...
+  do
+    {
+      result = lib_fwrite(buffer, len, sthis->stream);
+      if (result >= 0)
+        {
+          this->nput += result;
+          return result;
+        }
+
+      errcode = _NX_GETERRNO(result);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_stdsostream.c:92]: result = _NX_GETERRNO(result);
and remove errcode

Comment (Gary-Hobson) [reply libs/libc/stream/lib_stdsostream.c]: done

File: libs/libc/stream/lib_stdoutstream.c [hunk 17]
```diff
@@ -64,6 +65,43 @@ static void stdoutstream_putc(FAR struct lib_outstream_s *this, int ch)
...
+          return result;
+        }
+
+      result = _NX_GETERRNO(result);
+
+      /* EINTR (meaning that fputc was interrupted by a signal) is the only
+       * recoverable error.
+       */
+    }
+  while (result == EINTR);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_stdoutstream.c:100]:
```suggestion
  while (result == -EINTR);
```

File: libs/libc/stream/lib_stdsostream.c [hunk 18]
```diff
@@ -63,6 +63,42 @@ static void stdsostream_putc(FAR struct lib_sostream_s *this, int ch)
...
+  do
+    {
+      result = lib_fwrite(buffer, len, sthis->stream);
+      if (result >= 0)
+        {
+          this->nput += result;
+          return result;
+        }
+
+      result = _NX_GETERRNO(result);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_stdsostream.c:91]:
```suggestion
      result = _NX_GETERRVAL(result);
```

File: libs/libc/stream/lib_stdsostream.c [hunk 19]
```diff
@@ -63,6 +63,42 @@ static void stdsostream_putc(FAR struct lib_sostream_s *this, int ch)
...
+
+      /* EINTR (meaning that fputc was interrupted by a signal) is the only
+       * recoverable error.
+       */
+    }
+  while (result == EINTR);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_stdsostream.c:97]:
```suggestion
  while (result == -EINTR);
```

File: libs/libc/stream/lib_stdoutstream.c [hunk 20]
```diff
@@ -64,6 +65,43 @@ static void stdoutstream_putc(FAR struct lib_outstream_s *this, int ch)
...
+  do
+    {
+      result = fwrite(buffer, len, 1, sthis->stream);
+      if (result >= 0)
+        {
+          this->nput += result;
+          return result;
+        }
+
+      result = _NX_GETERRNO(result);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_stdoutstream.c:94]:
```suggestion
      result = _NX_GETERRVAL(result);
```

File: libs/libc/stream/lib_stdinstream.c [hunk 21]
```diff
@@ -52,6 +53,33 @@ static int stdinstream_getc(FAR struct lib_instream_s *this)
...
+  /* Get the buffer from the incoming stream */
+
+  nread = fread(buffer, len, 1, sthis->stream);
+  if (nread >= 0)
+    {
+      this->nget += nread;
+    }
+  else
+    {
+      nread = _NX_GETERRNO(nread);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_stdinstream.c:77]:
```suggestion
      nread = _NX_GETERRVAL(nread);
```

File: libs/libc/stream/lib_rawsostream.c [hunk 22]
```diff
@@ -73,6 +73,44 @@ static void rawsostream_putc(FAR struct lib_sostream_s *this, int ch)
...
+          this->nput += nwritten;
+          return nwritten;
+        }
+
+      /* The only expected error is EINTR, meaning that the write operation
+       * was awakened by a signal.  Zero would not be a valid return value
+       * from _NX_WRITE().
+       */
+
+      nwritten = _NX_GETERRNO(nwritten);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_rawsostream.c:106]:
```suggestion
      nwritten = _NX_GETERRVAL(nwritten);
```

File: libs/libc/stream/lib_rawsistream.c [hunk 23]
```diff
@@ -66,6 +66,33 @@ static int rawsistream_getc(FAR struct lib_sistream_s *this)
...
+  /* Attempt to read a buffer */
+
+  nread = _NX_READ(rthis->fd, buffer, len);
+  if (nread >= 0)
+    {
+      this->nget += nread;
+    }
+  else
+    {
+      nread = _NX_GETERRNO(nread);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_rawsistream.c:90]:
```suggestion
      nread = _NX_GETERRVAL(nread);
```

File: libs/libc/stream/lib_rawinstream.c [hunk 24]
```diff
@@ -66,6 +66,33 @@ static int rawinstream_getc(FAR struct lib_instream_s *this)
...
+  /* Attempt to read one character */
+
+  nread = _NX_READ(rthis->fd, buffer, len);
+  if (nread >= 0)
+    {
+      this->nget += nread;
+    }
+  else
+    {
+      nread = _NX_GETERRNO(nread);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_rawinstream.c:90]:
```suggestion
      nread = _NX_GETERRVAL(nread);
```

File: libs/libc/stream/lib_stdsostream.c (hunk 19 above)
Comment (Gary-Hobson) [reply libs/libc/stream/lib_stdsostream.c]: done

File: libs/libc/stream/lib_rawinstream.c (hunk 24 above)
Comment (Gary-Hobson) [reply libs/libc/stream/lib_rawinstream.c]: done

**Review by xiaoxiang781216 — APPROVED** [review]

File: libs/libc/stream/lib_stdsistream.c [hunk 25]
```diff
@@ -52,6 +52,29 @@ static int stdsistream_getc(FAR struct lib_sistream_s *this)
...
+
+  DEBUGASSERT(this);
+
+  /* Get the buffer from the incoming stream */
+
+  nread = fread(buffer, len, 1, sthis->stream);
+  if (nread >= 0)
+    {
+      this->nget += nread;
+    }
```
Comment (pkarashchenko) [inline libs/libc/stream/lib_stdsistream.c:73]: Don't we need 
```
  else
    {
      nread = _NX_GETERRVAL(nread);
    }
```
here as well?

Comment (xiaoxiang781216) [reply libs/libc/stream/lib_stdsistream.c:73]: Yes, @Gary-Hobson please fix it

**Pull request merged**